defrag/defrag.go

```diff
@@ -0,0 +1,276 @@
+}
+
+// FragmentList holds a container/list used to contains IP
+// packets/fragments.  It stores internal counters to track the
+// maximum total of byte, and the current length it has received.
+// It also stores a flag to know if he has seen the last packet.
+type FragmentList struct {
+	List          list.List
+	Total         uint16
+	Current       uint16
+	FinalReceived bool
+	LastSeen      time.Time
+}
+
+// Insert insert an IPv4 fragment/packet into the Fragment List
+// It use the following strategy : we are inserting fragment based
+// on their offset, latest first. This is sometimes called BSD-Right.
+// See: http://www.sans.org/reading-room/whitepapers/detection/ip-fragment-reassembly-scapy-33969
+func (f *FragmentList) Insert(ip *layers.IPv4) (bool, error) {
+	if ip.FragOffset*8 >= f.Total {
+		f.List.PushBack(ip)
+	} else {
+		for e := f.List.Front(); e != nil; e = e.Next() {
+			frag, _ := e.Value.(*layers.IPv4)
+			if ip.FragOffset <= frag.FragOffset {
+				debug.Printf("defrag: inserting frag %d before existing frag %d \n",
+					ip.FragOffset*8, frag.FragOffset*8)
+				f.List.InsertBefore(ip, e)
+				break
+			}
+		}
+	}
+	// packet.Metadata().Timestamp should have been better, but
+	// we don't have this info there...
+	f.LastSeen = time.Now()
+
+	fragLength := ip.Length - 20
+	// After inserting the Fragment, we update the counters
+	if f.Total < ip.FragOffset*8+fragLength {
+		f.Total = ip.FragOffset*8 + fragLength
+	}
+	f.Current = f.Current + fragLength
+
+	debug.Printf("defrag: insert ListLen: %d Total:%d Current:%d\n",
+		f.List.Len(),
+		f.Total, f.Current)
+
+	// Final Fragment ?
+	if ip.Flags&layers.IPv4MoreFragments == 0 {
+		f.FinalReceived = true
+	}
+	// Ready to try defrag ?
+	if f.FinalReceived && f.Total == f.Current {
+		return f.Build(ip)
+	}
+	return false, nil
+}
+
+// Build builds the final datagram, modifying ip in place.
+// It puts priority to packet in the early position of the list.
+// See Insert for more details.
+func (f *FragmentList) Build(ip *layers.IPv4) (bool, error) {
+	var final []byte
+	var currentOffset uint16 = 0
+
+	debug.Printf("defrag: building the datagram \n")
+	for e := f.List.Front(); e != nil; e = e.Next() {
+		frag, _ := e.Value.(*layers.IPv4)
+		if frag.FragOffset*8 == currentOffset {
+			debug.Printf("defrag: building - adding %d\n", frag.FragOffset*8)
+			final = append(final, frag.Payload...)
+			currentOffset = currentOffset + frag.Length - 20
+		} else if frag.FragOffset*8 < currentOffset {
+			// overlapping fragment - let's take only what we need
+			startAt := currentOffset - frag.FragOffset*8
+			debug.Printf("defrag: building - overlapping, starting at %d\n",
+				startAt)
+			if startAt > frag.Length-20 {
+				return false, fmt.Errorf("defrag: building - invalid fragment")
+			}
+			final = append(final, frag.Payload[startAt:]...)
+			currentOffset = currentOffset + frag.FragOffset*8
+		} else {
+			// Houston - we have an hole !
+			return false, nil
```

Comment:
Is this an implementation error?  Should we ever call Build unless we already have all the packets?